### PR TITLE
[#2016] Remove hash symbol from URL when decoding hash

### LINKED ISSUE #2016: Breakdown by file type not retained after refresh
### Please include the steps to reproduce the bug.

1. Open a report
2. Check 'breakdown by file type'
3. Check 'all'
4. Refresh

### What was expected to happen?
The checked files should be retained after refreshing.

### What actually happened? Please include a screenshot of the output.
After each refresh, the files become unchecked one by one

https://github.com/reposense/RepoSense/assets/97350904/2492c5f6-63b4-44af-bfc8-83ea6ac66f6f

### PR DESCRIPTION
**Pull request by jonasongg**
Fixes #2016 

### Proposed commit message

```
Fix bug where some params were not retained after refresh

The issue pertains to breakdown of file types disappearing after a
refresh, but it seems to apply to all params that were last in the URL
due to symbols not being filtered out correctly.

Let's fix this issue to make sure the state of the report stays the
same when refreshing.
```

### Other information

The bug seems to have been caused by the URL containing `#/` at the end, which would cause the last param to not be parsed correctly. I couldn't definitively find what caused the anchor to appear in the URL, so I filtered it out of the URL when decoding `hashParams`. Let me know if this behaviour should be changed!

Regarding the Cypress tests, #2017 seems to be outdated/some of the tests don't seem to be checking for the correct thing, at least on my machine. If I should investigate this further in anticipation of its being merged, let me know as well!

### CODE REVIEW

**Comment by sopa301** [comment]
Googling a bit suggests that this appending of `#/` is caused by Vue's [Hash Mode](https://router.vuejs.org/guide/essentials/history-mode), which appends a `#` followed by the route `/` (the default one) at the end of the URL. This could be a regression from the implementation of Vue Router.

I tested this using the 2324S1 CS2103T RepoSense dashboard. Some evidence for this hypothesis:

- Using `#/widget` gives a different view of the webpage.
- Using `#/test` gives a blank webpage

That said, the current fix is good enough to make it work. I propose 2 possible paths in the future:

1. Rewrite the code to ensure Vue Router's internal routing is finished and the URL is fixed before it is processed by `api.ts` (not recommended due to complexity)
2. Refactor `api.ts` to capture this consideration more distinctly.

**Comment by jonasongg** [comment]
@sopa301 Thanks for the suggestions! 

Just wanted to link these related issues here since it seems they are caused by this same problem: #2058 #1987

**Review by sopa301 — APPROVED** [review]
LGTM. As commented above, this is sufficient to fix the bug, but may warrant further refactoring.

**Review by ckcherry23 — COMMENTED** [review]
Thanks for your inputs @jonasongg and @sopa301. I think this temporary fix is a reasonable solution for now as it resolves the current issue without introducing new issues. We can switch to the Vue Router HTML5 mode in the future. 

@jonasongg Since this fix is somewhat unconventional, can you add a comment in the code explaining the rationale behind the change? This will help future developers understand the context and reasoning behind this decision.

Regarding the Cypress tests, we should ideally be adding a test case for the bug fixed in this PR itself. Let us wait to see the progress of PR #2017 over the next few days and make a call over the changes needed.

**Review by MarcusTXK — COMMENTED** [review]
Nice fix. Agree with @ckcherry23 that a comment to explain the logic for future developers would be useful.

Regarding the tests, I think perhaps @jonasongg could reference the code written in #2017 and add the updated tests, then co-author credit could be added to @nseah21.

**Comment by jonasongg** [comment]
@ckcherry23 @MarcusTXK Thanks for the reviews! Just to confirm, I should add the `codeView_renderFilterHash.cy.js` file from #2017 here and add tests that test this new expected behaviour?

**Comment by ckcherry23** [comment]
@jonasongg #2017 was just merged. You can look into what additions/changes are required on the tests now, thanks!

**Comment by MarcusTXK** [comment]
@jonasongg After discussion, we decided to merge in #2017, so you can work on the merged test directly on your branch.

**Comment by jonasongg** [comment]
Just added the comment and some test cases! For checked file types, I added some tests to make sure it was testing the last file type to appear in the URL (yml) as that was what was being affected by the bug. Also added some non exhaustive test cases for the code panel options including one for authorshipFileGlob that is most likely the cause for #2058 and #1987. Let me know if I should add more/change anything!

File: frontend/cypress/tests/codeView/codeView_renderFilterHash.cy.js [hunk 1]
```diff
@@ -309,31 +309,184 @@ describe('render filter hash', () => {
     cy.url()
       .should('not.contain', 'gradle');
 
+    cy.url()
+      .should('not.contain', 'yml');
+
```
Comment (ckcherry23) [inline frontend/cypress/tests/codeView/codeView_renderFilterHash.cy.js:314]: Can we make it clear why gradle and yml were the 2 file types selected here with a comment?
For example,
```
// Assumption: gradle is the first file type and yml is the last file type

File: frontend/cypress/tests/codeView/codeView_renderFilterHash.cy.js [hunk 2]
```diff
@@ -309,31 +309,184 @@ describe('render filter hash', () => {
...
+    cy.reload();
+
+    cy.url()
+      .should('not.contain', '%23%2F');
+
+    cy.url()
+      .should('contain', 'reverseAuthorshipOrder=true');
+  });
+
+  it('code panel: filter by glob: url params should persist after change and reload', () => {
```
Comment (ckcherry23) [inline frontend/cypress/tests/codeView/codeView_renderFilterHash.cy.js:461]: This function only tests after one reload, but there is a bug in the filter by glob string after 2 reloads. 
Hence, just wanted to add a note that we must add a test with 2 page reloads when the bug in #1987 gets fixed.

**Review by ckcherry23 — APPROVED** [review]
LGTM! Just have a few minor comments

**Comment by jonasongg** [comment]
Should be clarified now!

**Review by MarcusTXK — APPROVED** [review]
LGTM, I think you did a great job

**Pull request merged**